### ISSUE TITLE
hvd + tf2 eager mode is slow, allreduce step takes too much time

### ISSUE THREAD

**Issue by naturomics** [horovod/horovod#1628]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.0, 2.1
3. Horovod version:
4. MPI version:
5. CUDA version: 10.2
6. NCCL version: 2.5
7. Python version: 3.6
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
I'm using horovod for distributed training with tf2 `eager mode` (tf2 itself has distributed training support but is [buggy with eager mode](https://github.com/tensorflow/tensorflow/issues/30855)). And my model has custom gradients thus I have to make some changes to the hvd.DistributedOptimizer, see below, the changes are only merging `get_gradients` part into `apply_gradients` function. When using it in my code, the allreduce step takes a long time that a multi-GPUs setting has only the same speed as a single GPU setting, half of time GPUs are idle due to this step. Is there I missed something or horovod has not good support with tf eager mode yet? Any help is appreciated.

```
def create_distributed_optimizer(optimizer, name=None, device_dense='', device_sparse='',
                         compression=hvd.Compression.none, sparse_as_dense=False):
    class _DistributedOptimizer(tf.keras.optimizers.Optimizer):
        def __init__(self, name, device_dense, device_sparse, compression, sparse_as_dense, config):
            if name is None:
                name = "Distributed%s" % self.__class__.__base__.__name__
            self._name = name
            self._device_dense = device_dense
            self._device_sparse = device_sparse
            self._compression = compression
            self._sparse_as_dense = sparse_as_dense
            super(self.__class__, self).__init__(**config)

        def apply_gradients(self, grads_and_vars, *args, **kwargs):
            if hvd.size() > 1:
                averaged_grads, params = [], []
                with tf.name_scope(self._name + "_Allreduce"):
                    for grad, param in grads_and_vars:
                        if grad is not None:
                            if self._sparse_as_dense and isinstance(grad, tf.IndexedSlices):
                                grad = tf.convert_to_tensor(grad)
                            avg_grad = hvd.allreduce(grad,
                                                     device_dense=self._device_dense,
                                                     device_sparse=self._device_sparse,
                                                     compression=self._compression)
                            averaged_grads.append(avg_grad)
                        else:
                            averaged_grads.append(None)
                        params.append(param)
                grads_and_vars = zip(averaged_grads, params)
            return super(self.__class__, self).apply_gradients(grads_and_vars, *args, **kwargs)

        @classmethod
        def from_config(cls, cfg):
            return cls(name, device_dense, device_sparse, compression, sparse_as_dense, cfg)

    cls = type(optimizer.__class__.__name__, (optimizer.__class__,), dict(_DistributedOptimizer.__dict__))
    return cls(name, device_dense, device_sparse, compression, sparse_as_dense, optimizer.get_config())
```

Run in eager mode and model is not wrapped with tf.function:
```
hvd.init()
opt = tf.keres.optimizers.Adam()
opt = create_distributed_optimizer(opt)

dataset = tf.data.Dataset(...)
mode = Model()

first_batch = True
for input in dataset:
    loss = model(input)  # model is not wrapped with tf.function
   grads_and_vars = custom_gradient(loss)
   ops.apply_gradients(grads_and_vars)
    if first_batch:
        hvd.broadcast_variables(model.variables, root_rank=0)
        hvd.broadcast_variables(opt.variables(), root_rank=0)
        first_batch = False
```

**Comment by naturomics** [comment]
Closing. This implementation has better performance:

```
def create_distributed_optimizer(optimizer, name=None, device_dense='', device_sparse='',
                         compression=hvd.Compression.none, sparse_as_dense=False):
    class _DistributedOptimizer(tf.keras.optimizers.Optimizer):
        def __init__(self, name, device_dense, device_sparse, compression, sparse_as_dense, config):
            if name is None:
                name = "Distributed%s" % self.__class__.__base__.__name__
            self._allreduce_grads = hvd._make_allreduce_grads_fn(
                name, device_dense, device_sparse, compression, sparse_as_dense)
            super(self.__class__, self).__init__(**config)

        def apply_gradients(self, grads_and_vars, *args, **kwargs):
            if hvd.size() > 1:
                grads, vars = zip(*grads_and_vars)
                avg_grads = self._allreduce_grads(grads)
                grads_and_vars = list(zip(avg_grads, vars))
            return super(self.__class__, self).apply_gradients(grads_and_vars, *args, **kwargs)

        @classmethod
        def from_config(cls, cfg):
            return cls(name, device_dense, device_sparse, compression, sparse_as_dense, cfg)

    cls = type(optimizer.__class__.__name__, (optimizer.__class__,), dict(_DistributedOptimizer.__dict__))
    return cls(name, device_dense, device_sparse, compression, sparse_as_dense, optimizer.get_config())
```
